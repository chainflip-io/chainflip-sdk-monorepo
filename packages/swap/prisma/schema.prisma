// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

// `pnpm prisma format` after changes :)

generator client {
  provider        = "prisma-client-js"
  // https://github.com/prisma/prisma/issues/2584#issuecomment-646414549
  // https://github.com/prisma/prisma/issues/6603#issuecomment-849979849
  output          = "../node_modules/.prisma/client"
  previewFeatures = ["multiSchema"]
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
  schemas  = ["public", "private"]
}

enum Chain {
  Polkadot
  Ethereum
  Bitcoin

  @@schema("public")
}

enum Asset {
  FLIP
  USDC
  DOT
  ETH
  BTC

  @@schema("public")
}

model SwapDepositChannel {
  id                    BigInt   @id @default(autoincrement())
  uuid                  String   @unique @default(uuid())
  srcAsset              Asset
  depositAddress        String
  expectedDepositAmount Decimal  @db.Decimal(30, 0)
  destAsset             Asset
  destAddress           String
  expiryBlock           Int
  issuedBlock           Int
  createdAt             DateTime @default(now())
  swaps                 Swap[]

  @@index([depositAddress])
  @@schema("public")
}

model Swap {
  id                        BigInt              @id @default(autoincrement())
  /// the swap id on the state chain
  nativeId                  BigInt              @unique
  srcAsset                  Asset
  depositAmount             Decimal              @db.Decimal(30, 0)
  intermediateAmount        Decimal?             @db.Decimal(30, 0)
  egressAmount              Decimal?             @db.Decimal(30, 0)
  destAsset                 Asset
  destAddress               String
  depositReceivedAt         DateTime
  depositReceivedBlockIndex String
  swapExecutedAt            DateTime?
  swapExecutedBlockIndex    String?
  egressCompletedAt         DateTime?
  egressCompletedBlockIndex String?
  swapDepositChannel        SwapDepositChannel? @relation(fields: [swapDepositChannelId], references: [id])
  swapDepositChannelId      BigInt?
  txHash                    String?             @unique
  egress                    Egress?
  createdAt                 DateTime            @default(now())
  updatedAt                 DateTime            @default(now()) @updatedAt

  @@schema("public")
}

model Egress {
  id          BigInt     @id @default(autoincrement())
  /// the egress id on the state chain
  nativeId    BigInt
  /// the swap id on the state chain
  chain       Chain
  amount      Decimal    @db.Decimal(30, 0)
  timestamp   DateTime
  swap        Swap?      @relation(fields: [swapId], references: [id])
  swapId      BigInt?    @unique
  broadcast   Broadcast? @relation(fields: [broadcastId], references: [id])
  broadcastId BigInt?
  createdAt   DateTime   @default(now())

  @@unique([nativeId, chain])
  @@schema("public")
}

model Broadcast {
  id       BigInt   @id @default(autoincrement())
  /// the broadcast id on the state chain
  nativeId BigInt
  chain    Chain
  egresses Egress[]

  @@unique([nativeId, chain])
  @@schema("public")
}

model State {
  id     Int @id @default(autoincrement())
  height Int @default(0)

  @@schema("private")
}

model MarketMaker {
  id        Int    @id @default(autoincrement())
  name      String @unique
  publicKey String

  @@schema("private")
}

enum ThirdPartyProtocol {
  lifi
  squid

  @@schema("public")
}

model ThirdPartySwap {
  id            BigInt             @id @default(autoincrement())
  uuid          String             @unique
  protocol      ThirdPartyProtocol
  routeResponse Json
  txHash        String
  txLink        String
  createdAt     DateTime           @default(now())

  @@schema("public")
}
