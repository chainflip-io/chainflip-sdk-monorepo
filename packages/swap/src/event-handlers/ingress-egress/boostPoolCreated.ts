import { arbitrumIngressEgressBoostPoolCreated } from '@chainflip/processor/141/arbitrumIngressEgress/boostPoolCreated';
import { bitcoinIngressEgressBoostPoolCreated } from '@chainflip/processor/141/bitcoinIngressEgress/boostPoolCreated';
import { ethereumIngressEgressBoostPoolCreated } from '@chainflip/processor/141/ethereumIngressEgress/boostPoolCreated';
import { polkadotIngressEgressBoostPoolCreated } from '@chainflip/processor/141/polkadotIngressEgress/boostPoolCreated';
import { solanaIngressEgressBoostPoolCreated } from '@chainflip/processor/160/solanaIngressEgress/boostPoolCreated';
import { assethubIngressEgressBoostPoolCreated } from '@chainflip/processor/190/assethubIngressEgress/boostPoolCreated';
import { ChainflipChain } from '@chainflip/utils/chainflip';
import { z } from 'zod';
import { EventHandlerArgs } from '..';

const schemas = {
  Arbitrum: arbitrumIngressEgressBoostPoolCreated,
  Bitcoin: bitcoinIngressEgressBoostPoolCreated,
  Ethereum: ethereumIngressEgressBoostPoolCreated,
  Polkadot: polkadotIngressEgressBoostPoolCreated,
  Solana: solanaIngressEgressBoostPoolCreated,
  Assethub: assethubIngressEgressBoostPoolCreated,
} as const satisfies Record<ChainflipChain, z.ZodTypeAny>;

export const boostPoolCreated =
  (chain: ChainflipChain) =>
  async ({ prisma, event }: EventHandlerArgs) => {
    const {
      boostPool: { asset, tier },
    } = schemas[chain].parse(event.args);

    await prisma.boostPool.upsert({
      create: {
        asset,
        feeTierPips: tier,
        // safe mode is disabled by default which means pools are active
        boostEnabled: true,
        depositEnabled: true,
        withdrawEnabled: true,
      },
      where: {
        asset_feeTierPips: {
          asset,
          feeTierPips: tier,
        },
      },
      update: {},
    });
  };
