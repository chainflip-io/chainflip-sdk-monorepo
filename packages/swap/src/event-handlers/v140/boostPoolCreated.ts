import { z } from 'zod';
import { assetAndChain, number } from '@/shared/parsers';
import { screamingSnakeToPascalCase } from '@/shared/strings';
import { EventHandlerArgs } from '..';
import { InternalAsset } from '../../client';

const boostPoolCreatedEventSchema = z.object({
  boostPool: z.object({
    asset: assetAndChain.transform((chainAsset) => ({
      ...chainAsset,
      asset: screamingSnakeToPascalCase(chainAsset.asset),
    })),
    tier: number,
  }),
});

export const boostPoolCreated = async ({ prisma, event }: EventHandlerArgs) => {
  const {
    boostPool: { asset: chainAsset, tier },
  } = boostPoolCreatedEventSchema.parse(event.args);

  await prisma.boostPool.upsert({
    create: {
      asset: chainAsset.asset as InternalAsset,
      feeTierPips: tier,
      // safe mode is disabled by default which means pools are active
      boostEnabled: true,
      depositEnabled: true,
      withdrawEnabled: true,
    },
    where: {
      asset_feeTierPips: {
        asset: chainAsset.asset as InternalAsset,
        feeTierPips: tier,
      },
    },
    update: {},
  });
};
