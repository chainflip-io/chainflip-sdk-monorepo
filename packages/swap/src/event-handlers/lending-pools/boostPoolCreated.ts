import { lendingPoolsBoostPoolCreated as schema } from '@chainflip/processor/11000/lendingPools/boostPoolCreated';
import { EventHandlerArgs } from '../index.js';

export const lendingPoolsBoostPoolCreated = async ({ prisma, event }: EventHandlerArgs) => {
  const {
    boostPool: { asset, tier },
  } = schema.parse(event.args);

  await prisma.boostPool.upsert({
    create: {
      asset,
      feeTierPips: tier,
      // safe mode is disabled by default which means pools are active
      boostEnabled: true,
      depositEnabled: true,
      withdrawEnabled: true,
    },
    where: {
      asset_feeTierPips: {
        asset,
        feeTierPips: tier,
      },
    },
    update: {},
  });
};
