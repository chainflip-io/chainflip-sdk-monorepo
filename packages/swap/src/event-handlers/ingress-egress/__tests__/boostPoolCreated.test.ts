import { describe, it, expect, beforeEach } from 'vitest';
import prisma from '../../../client';
import { boostPoolCreated } from '../boostPoolCreated';

describe(boostPoolCreated, () => {
  beforeEach(async () => {
    await prisma.$queryRaw`TRUNCATE TABLE "BoostPool" CASCADE`;
  });
  it('handles event by creating a boost pool entry', async () => {
    await prisma.$transaction(async (txClient) => {
      await boostPoolCreated('Bitcoin')({
        prisma: txClient,
        block: {
          height: 120,
          timestamp: 1670337105000,
        } as any,
        event: {
          args: {
            boostPool: {
              asset: {
                __kind: 'Btc',
              },
              tier: 10,
            },
          },
          name: 'BitcoinIngressEgress.BoostPoolCreated',
          indexInBlock: 7,
        },
      });
    });

    expect(await prisma.boostPool.findFirst()).toMatchSnapshot({
      id: expect.any(Number),
    });
  });
});
