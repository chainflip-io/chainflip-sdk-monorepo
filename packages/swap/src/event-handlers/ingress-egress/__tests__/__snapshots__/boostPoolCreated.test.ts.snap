// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`boostPoolCreated > handles event by creating a boost pool entry 1`] = `
{
  "asset": "Btc",
  "boostEnabled": true,
  "depositEnabled": true,
  "feeTierPips": 10,
  "id": Any<Number>,
  "withdrawEnabled": true,
}
`;
